piquasso/dual_rail_encoding.py

```diff
@@ -0,0 +1,292 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+r"""Dual-rail encoding utilities for bosonic qubits.
+
+This module provides functions to encode qubit circuits defined in Qiskit into dual-rail
+bosonic qubit Piquasso programs.
+
+For the dual-rail encoding, the following convention is used:
+
+.. math::
+    |0\rangle_{\text{qubit}} = \ket{0, 1}_{\text{qumodes}} \\
+    |1\rangle_{\text{qubit}} = \ket{1, 0}_{\text{qumodes}}
+"""
+
+import piquasso as pq
+import numpy as np
+
+zero_bosonic_qubit_state = [0, 1]
+one_bosonic_qubit_state = [1, 0]
+
+
+def prep_bosonic_qubits(all_modes: list, modes_with_one_photon: list) -> list:
+    r"""Prepares a bosonic qubits in the specified basis states.
+
+    The following dual-rail encoding convention is being used:
+
+    .. math::
+          |0\rangle_{\text{qubit}} = \ket{0, 1}_{\text{qumodes}} \\
+          |1\rangle_{\text{qubit}} = \ket{1, 0}_{\text{qumodes}}
+
+    Args:
+        all_modes: All the modes to first prepare in the vacuum state.
+        modes_with_one_photon: The modes where a single photon is created.
+
+    Returns:
+        A list of Piquasso instructions to prepare the bosonic qubit.
+    """
+    instructions = []
+    if all_modes:
+        instructions.append(pq.Vacuum().on_modes(*all_modes))
+    if modes_with_one_photon:
+        instructions.append(pq.Create().on_modes(*modes_with_one_photon))
+    return instructions
+
+
+def paulix_bosonic(mode1, mode2):
+    """Applies a Pauli-X gate on a bosonic qubit encoded in dual-rail format.
+
+    Args:
+        mode: The mode of the bosonic qubit.
+    """
+    instructions = []
+    instructions.append(pq.Beamsplitter(np.pi / 2).on_modes(mode1, mode2))
+    instructions.append(pq.Phaseshifter(np.pi).on_modes(mode1))
+    return instructions
+
+
+def pauliz_bosonic(mode1, mode2):
+    """Applies a Pauli-Z gate on a bosonic qubit encoded in dual-rail format.
+
+    Args:
+        mode: The mode of the bosonic qubit.
+    """
+    return phase_gate_bosonic(np.pi, mode1)
+
+
+def phase_gate_bosonic(theta, mode):
+    """Applies a phase gate on a bosonic qubit encoded in dual-rail format.
+
+    Args:
+        theta: The phase angle.
+        mode: The first mode of the bosonic qubit (the second mode is mode + 1).
+    Returns:
+        A list of Piquasso instructions implementing the phase gate.
+    """
+    instructions = []
+    if not np.isclose(theta, 0.0):
+        instructions.append(pq.Phaseshifter(theta).on_modes(mode))
+    return instructions
+
+
+def hadamard_bosonic(mode1, mode2):

```

Comment:
If it is public, then it would need a docstring (similarly for `pauliz_bosonic`)